### PR TITLE
GRW-925 / Feature / Implement quote cart session

### PR DESCRIPTION
**Pull request by robinandeer**
## What?

Implement new quote cart session as Koa middleware.
Remove createQuoteCart hook. Replace with onboardingQuoteCartId hook.
Update Debugger and Embark to use new quote cart session.

## Why?

We want to implement a way to share a quote cart between different pages like forever and embark.

This sounds very much like a session so let's treat it as such and store a quote cart ID in a cookie available to the whole app.

The quote cart is checked every time the app boots up and recreated if it no longer exists or isn't active (checkout started).

When on the offer page we get the quote cart from the URL instead.

**Ticket(s): [GRW-925]**

[GRW-925]: https://hedvig.atlassian.net/browse/GRW-925?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/server/middleware/quoteCart.ts [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+              id
+            }
+          }
+        `,
+      variables: { market: apiMarket, locale: isoLocale },
+    })
+
+    if (result.status !== 200) {
+      throw new Error(`Unable to create new quote cart ${result.status}`)
+    }
```
Comment (robinandeer) [inline src/server/middleware/quoteCart.ts:95]: Check if we get a quote cart id instead

File: src/server/middleware/quoteCart.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+            }
+          }
+        }
+      `,
+      variables: { id },
+    })
+
+    if (response.status !== 200) {
+      throw new Error(`Giraffe returned status ${response.status}`)
+    }
```
Comment (robinandeer) [inline src/server/middleware/quoteCart.ts:48]: Is this needed?

File: src/server/middleware/quoteCart.ts [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+  const storage = new ServerCookieStorage(ctx)
+  const quoteCartId = storage.getItem(ONBOARDING_QUOTE_CART_COOKIE_KEY)
+
+  const performReset = ctx.query.reset === 'true'
+  if (
+    !performReset &&
+    quoteCartId &&
+    (await isQuoteCartActive(quoteCartId, apiMarket))
+  ) {
```
Comment (robinandeer) [inline src/server/middleware/quoteCart.ts:74]:
```suggestion
  const performReset = ctx.query.reset === 'true'
  const shouldNotCreateCart = performReset === false && quoteCartId && (await isQuoteCartActive(quoteCartId, apiMarket))
  
  if (shouldNotCreateCart) {
```

File: src/client/utils/hooks/useOnboardingQuoteCartId.ts [hunk 4]
```diff
@@ -0,0 +1,11 @@
+import { CookieStorage } from 'cookie-storage'
+import { ONBOARDING_QUOTE_CART_COOKIE_KEY } from 'shared/sessionStorage'
+
+export const useOnboardingQuoteCartId = () => {
+  const cookieStorage = new CookieStorage()
```
Comment (robinandeer) [inline src/client/utils/hooks/useOnboardingQuoteCartId.ts:5]: move outside hook?

File: src/server/middleware/quoteCart.ts [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+      variables: { market: apiMarket, locale: isoLocale },
+    })
+
+    if (result.status !== 200) {
+      throw new Error(`Unable to create new quote cart ${result.status}`)
+    }
+
+    const quoteCartId = result.data.data.onboardingQuoteCart_create.id
+    ctx.state.getLogger('quoteCart').info(`Created quote cart ${quoteCartId}`)
+    storage.setItem(ONBOARDING_QUOTE_CART_COOKIE_KEY, quoteCartId)
```
[the remark this answers is not in the document]
Comment (robinandeer) [reply src/server/middleware/quoteCart.ts:99]: Yes!

[the remark this answers is not in the document]
Comment (simonauner) [reply src/server/middleware/quoteCart.ts]: Secure by default! 👏

File: src/server/middleware/quoteCart.ts [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+  const { apiMarket, isoLocale } = locales[locale]
+
+  const storage = new ServerCookieStorage(ctx)
+  const quoteCartId = storage.getItem(ONBOARDING_QUOTE_CART_COOKIE_KEY)
+
+  const forceReset = ctx.query.reset === 'true'
+  const quoteCartIsUsable =
+    quoteCartId && (await isQuoteCartUsable(quoteCartId, apiMarket))
+
+  if (!forceReset && quoteCartIsUsable) {
```
Comment (simonauner) [inline src/server/middleware/quoteCart.ts:69]: Nice! Reading this makes sense :) 👍

**Review by simonauner — APPROVED** [review]

**Review by guilhermespopolin — APPROVED** [review]

**Pull request merged**